File: src/main/java/com/ploy/bubble_server_v3/common/jwt/Jwt.java
package com.ploy.bubble_server_v3.common.jwt;

import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.Claim;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.ploy.bubble_server_v3.common.jwt.dto.TokenResponse;
import com.ploy.bubble_server_v3.domain.user.domain.vo.Role;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.Arrays;
import java.util.Date;

public class Jwt {
    private final String issuer;
    private final Long tokenExpire;
    private final Long refreshTokenExpire;
    private final Algorithm algorithm;
    private final JWTVerifier jwtVerifier;

    public Jwt(String clientSecret, String issuer, Long tokenExpire, Long refreshTokenExpire) {
        this.issuer = issuer;
        this.tokenExpire = tokenExpire;
        this.refreshTokenExpire = refreshTokenExpire;
        this.algorithm = Algorithm.HMAC512(clientSecret);
        this.jwtVerifier = JWT.require(algorithm)
                .withIssuer(issuer).build();
    }

    public String generateAccessToken(Claims claims) {
        return sign(claims, tokenExpire);
    }

    public String generateRefreshToken(Claims claims) {
        return sign(claims, refreshTokenExpire);
    }

    public TokenResponse generateAllToken(Claims claims) {
        return new TokenResponse(generateAccessToken(claims), generateRefreshToken(claims));
    }

    private String sign(Claims claims, Long expireTime) {
        Date now = new Date();
        return JWT.create()
                .withIssuer(issuer)
                .withIssuedAt(now)
                .withExpiresAt(new Date(now.getTime() + expireTime * 1000L))
                .withClaim("userId", claims.userId)
                .withArrayClaim("roles", Arrays.stream(claims.roles)  // roles를 문자열 배열로 변환
                        .map(Role::name)  // Role 객체에서 이름(String)으로 변환
                        .toArray(String[]::new))  // String[] 배열로 변환
                .sign(algorithm);
    }

    public Claims verify(String token) {
        return new Claims(jwtVerifier.verify(token));
    }

    @Getter
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Claims {
        Long userId;
        Role[] roles;
        Date iat;
        Date exp;

        Claims(DecodedJWT decodedJwt) {
            Claim memberId = decodedJwt.getClaim("userId");
            if (!memberId.isNull()) {
                this.userId = memberId.asLong();
            }
            Claim roles = decodedJwt.getClaim("roles");
            if (!roles.isNull()) {
                this.roles = Arrays.stream(roles.asArray(String.class)) // 역할을 String 배열로 받아오기
                        .map(Role::valueOf) // String을 Role 객체로 변환
                        .toArray(Role[]::new); // Role[] 배열로 변환
            }
            this.iat = decodedJwt.getIssuedAt();
            this.exp = decodedJwt.getExpiresAt();
        }

        public static Claims from(Long memberId, Role[] roles) {
            Claims claims = new Claims();
            claims.userId = memberId;
            claims.roles = roles;
            return claims;
        }

        @Override
        public String toString() {
            return "Claims{"
                    + "userId=" + userId
                    + ", roles="
                    + Arrays.toString(roles)
                    + ", iat=" + iat
                    + ", exp="
                    + exp
                    + '}';
        }
    }
}

File: src/main/kotlin/com/ploy/bubble_server_v2/common/jwt/Jwt.kt
package com.ploy.bubble_server_v2.common.jwt

import com.auth0.jwt.JWT
import com.auth0.jwt.JWTVerifier
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.interfaces.DecodedJWT
import com.ploy.bubble_server_v2.common.dto.TokenResponse
import java.util.Date

class Jwt(
    clientSecret: String,
    private val issuer: String,
    private val tokenExpire: Int,
    private val refreshTokenExpire: Int
) {
    private val algorithm: Algorithm = Algorithm.HMAC512(clientSecret)
    private val jwtVerifier: JWTVerifier = JWT.require(algorithm)
        .withIssuer(issuer)
        .build()

    // 액세스 토큰을 생성하는 함수
    fun generateAccessToken(claims: Claims): String {
        return sign(claims, tokenExpire)
    }

    // 리프레시 토큰을 생성하는 함수
    fun generateRefreshToken(claims: Claims): String {
        return sign(claims, refreshTokenExpire)
    }

    // 액세스 토큰과 리프레시 토큰을 모두 생성하는 함수
    fun generateAllToken(claims: Claims): TokenResponse {
        return TokenResponse(generateAccessToken(claims), generateRefreshToken(claims))
    }

    // 토큰 서명 함수
    private fun sign(claims: Claims, expireTime: Int): String {
        val now = Date() // 현재 시간
        return JWT.create()
            .withIssuer(issuer)
            .withIssuedAt(now)
            .withExpiresAt(Date(now.time + expireTime * 1000L))
            .withClaim("memberId", claims.memberId)
            .sign(algorithm)
    }

    // 토큰을 검증하는 함수
    fun verify(token: String): Claims {
        return Claims(jwtVerifier.verify(token))
    }

    // Claims 데이터 클래스
    data class Claims(
        val memberId: Long?,
        val iat: Date?,
        val exp: Date?
    ) {
        constructor(decodedJwt: DecodedJWT) : this(
            memberId = decodedJwt.getClaim("memberId").takeIf { !it.isNull }?.asLong(),
            iat = decodedJwt.issuedAt,
            exp = decodedJwt.expiresAt
        )

        companion object {
            // 주어진 memberId로 Claims 생성
            fun from(memberId: Long): Claims {
                return Claims(memberId, null, null)
            }
        }

        override fun toString(): String {
            return "Claims(memberId=$memberId, iat=$iat, exp=$exp)"
        }
    }
}
